### PR TITLE
Engine: fix bug that allowed non-storable inputs to be passed to process

### LINKED ISSUE #5128: Process functions allow non-Data arguments to be passed as input
MWE:
```python
In [1]: from aiida import engine, orm

In [2]: @engine.calcfunction
   ...: def test_kwargs(**kwargs):
   ...:     for value in kwargs.values():
   ...:         assert isinstance(value, orm.Data)
   ...: 
   ...: 

In [3]: test_kwargs(**{'a': orm.Int(1)})
Out[3]: {}

In [4]: test_kwargs(**{'a': orm.Int(1), 'b': 1})
Report: [487|test_kwargs|on_except]: Traceback (most recent call last):
  File "/home/sph/.virtualenvs/aiida_dev/lib/python3.9/site-packages/plumpy/process_states.py", line 231, in execute
    result = self.run_fn(*self.args, **self.kwargs)
  File "/home/sph/code/aiida/env/dev/aiida-core/aiida/engine/processes/functions.py", line 395, in run
    result = self._func(*args, **kwargs)
  File "<ipython-input-2-5fb70fcb7652>", line 4, in test_kwargs
    assert isinstance(value, orm.Data)
AssertionError

```
As you can see, the second invocation of the `test_kwargs` calcfunction passes a normal `int` as an argument, but the input validation does not catch it and simply executes the function where the kwargs contains the normal `int` value. In contrast, if we don't use `**kwargs` but explicitly define the arguments, the validation _does_ work:
```python
In [5]: @engine.calcfunction
   ...: def test_explicit(a, b):
   ...:     assert isinstance(a, orm.Data)
   ...:     assert isinstance(b, orm.Data)

In [6]: test_explicit(a=orm.Int(1), b=5)

ValueError: Error occurred validating port 'inputs.b': value 'b' is not of the right type. Got '<class 'int'>', expected '(<class 'aiida.orm.nodes.data.data.Data'>,)'
```

Originally reported on the mailing list: https://groups.google.com/g/aiidausers/c/BmDOzte4vWQ/m/v7PZUTywBAAJ

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #5128 

The basic assumption for a `Process` in `aiida-core` is that all of its
inputs should be storable in the database as nodes. Under the current
link model, this means that they should be instances of the `Data` class
or subclasses thereof. There is a noticeable exception for ports that
are explicitly marked as `non_db=True`, in which case the value is not
linked as a node, but is stored as an attribute directly on the process
node itself, or not stored whatsoever.

This basic rule was never explicitly enforced, which made it possible to
define processes that would happily take non-storable inputs. The input
would not get stored in the database, but would be available within the
processes lifetimes from the `inputs` property allowing it to be used.
This will most likely result into unintentional loss of provenance.

The reason is that the default `valid_type` of the top-level inputs
namespace of the `Process` class was never being set to `Data`. This
meant that any type would be excepted for a `Process` and all its
subclasses unless the valid type of a port was explicitly overridden.
This meant that for normal dynamic namespaces, even non-storable types
would be accepted just fine. Setting `valid_type=Data` for the input
namespace of the `Process` class fixes the problem therefore.

### CODE REVIEW

File: tests/engine/processes/test_builder.py
Comment (ltalirz) [inline tests/engine/processes/test_builder.py:61]: Hmyeah, interesting that there were even tests that relied on this behavior ;-)

**Review by ltalirz — COMMENTED** [review]
Thanks @sphuber , this looks fine to me.

Just for me to understand: this only affected namespaces and ports therein or also input ports defined at the top level?

If the latter, then I could imagine some plugins already relying on this.

Let me know if you want me to approve or rather wait for the review by Marnik.

**Comment by sphuber** [comment]
> Just for me to understand: this only affected sub-namespaces and ports therein or also input ports defined at the top level?
> If the latter, then I could imagine some plugins already relying on this.

This affected the top-level input namespace so this includes all sub-namespaces and input ports at the top-level. If a plugin indeed relied on specifying an input port without a `Data` (or subclass thereof) as `valid_type`, then that would have "worked" even if `non_db` was not set to `True`. However, maybe it would be good to have that break, because they would probably unintentionally be losing provenance since the inputs are not actually stored in the database.

**Review by ltalirz — APPROVED** [review]

**Comment by ltalirz** [comment]
> even if `non_db` was not set to `True`

I guess you mean `False`.

I agree that this is a pattern we would like developers to get away from; just saying that this may break plugins. That said, a quick search on Github for strings like `valid_type=int`, `valid_type=str` did not turn up any results.

**Pull request merged**
